jooq/src/main/java/com/cerner/beadledom/jooq/JooqTransactionalHooks.java

```diff
@@ -0,0 +1,55 @@
+package com.cerner.beadledom.jooq;
+
+/**
+ * Transaction callback hooks to be executed when the current Jooq transaction completes.
+ *
+ * @author John Leacox
+ * @since 3.3
+ */
+public abstract class JooqTransactionalHooks {
+  /**
+   * Registers an action to be executed when the current Jooq transaction is successfully committed.
+   *
+   * <p>
+   * If this method is called while a Jooq transaction is not active, the action will be executed
+   * immediately.
+   * </p>
+   *
+   * <p>
+   * If the current Jooq transaction is rolled back instead of committed, then the action will be
+   * discarded and never called.
+   * </p>
+   *
+   * <h3>Savepoints / Nested Transactions</h3>
+   *
+   * <p>
+   * Actions registered on nested transaction (or savepoints) will be called after the top-level
+   * transaction is committed, but not if a rollback to the savepoint or any parent savepoint
+   * occurred during the transaction. When a savepoint is rolled back the commit hooks associated
+   * with that savepoint or further nested savepoints will not be executed. However, commit hooks
+   * associated with savepoints at a higher level of nesting than the rollback, will be executed if
+   * the top-level transaction commits.
+   * </p>
+   *
+   * <h3>Order of Execution</h3>
+   *
+   * <p>The callback actions will be executed in the order they are registered.</p>
+   *
+   * <h3>Exception Handling</h3>
+   *
+   * <p>
+   * If a callback action throws an exception, no later registered callbacks in the same
+   * transaction will be executed.
+   * </p>
+   *
+   * @param action the action to be executed upon the successful commit of the current Jooq
+   *     transaction
+   */
+  public abstract void whenCommitted(Runnable action);
+
+  void setTransaction(JooqTransaction transaction) {
```

Comment:
Perhaps this is subjective, but i always think twice before having multiple implementations within an inheritance hierarchy like you have for the setter/getter  here and in the `ThreadLocalJooqTransactionalHooks` class. Why do we need these methods in two places?

Also, it seems like this could be avoided if we make this an interface and have the setter/getter as default methods because this abstract class does not have state associated to it.

Reply:
The goal of how I split this up was to have a public facing API with only the `whenCommitted` method and then internal-only APIs necessary for the framework.

I think I tried this once before, but I want to try it again: what might work is to make this an interface with only `whenCommitted`, move `setTransaction` and `clearTransaction` to the implementation only, and Update `ThreadLocalCommitHookExecutingTransactionListener` to depend directly on the implementation. If we had multiple implementations we might need a different structure, but with only one implementation currently, if this works I think it should be fine.

Reply:
Ahh, one disadvantage of this is that we'll need a Guice binding on the implementation which exposes it beyond just this package.

Reply:
Looks like we already bind some package-private classes in Guice.

Reply:
I am always wary of any tool that uses reflections so heavily (Heard some horror stories about powermock)

Reply:
Changed to an interface: https://github.com/cerner/beadledom/pull/140/commits/ed118b69f588ab7f1fe9de2e975b4236cdd2e7a2